src/java-opentimelineio/src/main/java/io/opentimeline/opentimelineio/AnyVector.java

```diff
@@ -0,0 +1,100 @@
+package io.opentimeline.opentimelineio;
+
+import io.opentimeline.OTIONative;
+
+public class AnyVector extends OTIONative {
```

Comment:
For my undertstanding, how are the Dictionary and Vector objects used in C++? Just wondering if we could use regular Java collections, and if that would be a major refactor or not.

Reply:
Dictionary and Vector in C++ are similar to Map and List in java. 
I thought this would be more user friendly but AnyDictionary and AnyVector in the C++ implementation are derived from std::map and std::vector and in they additionally record a time-stamp that bumps monotonically everytime an operation is performed, like insert, clear, etc.

This was done to help with handing out mutation aware iterators in python and to prevent modification of the list while iterating. This MutationStamp also lets users know if the map was destroyed or relocated in memory.
I wasn't sure at the time if the java implementation would be similar to how it was done with python and if it would be necessary, so I just wrote a simple wrapper.

https://github.com/PixarAnimationStudios/OpenTimelineIO/blob/master/src/py-opentimelineio/opentimelineio-bindings/otio_anyDictionary.h
This is the python binding for AnySictionary.

Java doesn't allow modification of lists while iterating, so maybe we can use regular Java collections.

In case we go ahead with using regular Java collections, we can just scrap these classes and write utility functions for converting the Java collections to C++. I'll also have to replace these implementations in all other classes that use AnyDictionary or AnyVector.

Reply:
There are different options in Java that should allow for concurrent modification, good examples are[CopyOnWriteArrayList](https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/CopyOnWriteArrayList.html) and [ConcurrentHashMap](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ConcurrentHashMap.html). I would not recommend wrapping generic collections of C++ STD library just for convenience and fully lean on Java's native collections. I can guide you for the effect you want to achieve if not familiar.

Reply:
Just curious what the strategy is for the following:
  (1) You get a SerializableObject's metadata dictionary, it comes back as some Java Map class.
  (2) The dictionary is mutated (under the hood) without Java knowing specifically about it.
  (3) In Java you peer into that Map class instance.

Does Java rebuild it?  Does it know it's outdated?  etc.  In Python/Swift we make use of the fact that we can just write some methods that conform to a "protocol" and if we faithfully implement those we get something that behaves perfectly like a Python or Swift collection, and that perfectly tracks the C++ dictionary, because we didn't actually create a mirror of the collecction in Python or Swift.  Is the approach similar for Java?

Reply:
The current implementation just holds the reference to the c++ AnyDictionary. So I think any changes under the hood should reflect. 

But if we change the implementation to create a Java Map each time someone fetches metadata it definitely wouldn't reflect the changes. 

I'll try to read through and understand the approach in python/swift. 

Java interfaces are similar to Swift protocols.

Reply:
Pardon my ignorance, but are Vector, Dictionary used in OTIO context or used as a plain collection? Is there a way where we don't have any native handles on those, but only on the values inside Vector and Dictionary?

If you always have to rebuild a map from C++ from native objects, unless you already have a corresponding Java object you might need to generate a collection + for each value object in a collection the OTIONative counterpart. If that was the case there are improvements to make where C++ can diff what is inside the collection and do a "merge" but it can get quite involved. Possibly other ways to do this, but hoping to avoid the complicated merges (or unnecessarily heap allocation) and yet still use Java collections where possible.

Reply:
I think you do not want to change the implementation to create a Java Map then.  You should treat the C++ dictionary/vector as your "source of truth" and simply write a Java class that queries that source of truth as needed.  Anything else (i.e. trying to keep them in sync and do diffs etc.) lies in the path of madness.

Reply:
I was hoping that like Python and Swift, there was some sort of interface/protocol specification in Java, where, if you implement a small number of key methods, you get something that behaves algorithmically like the collection you want without actually having to make a copy of the data.  That is I believe what you need to aim for.

Reply:
@davidbaraff are you referring to the methods like `get_item` , `set_item` here:
https://github.com/PixarAnimationStudios/OpenTimelineIO/blob/master/src/py-opentimelineio/opentimelineio-bindings/otio_anyDictionary.h

Reply:
Yes.  In Python, you can make something that feels like a Python dictionary by implement about 5 key methods (get, set, delete, iter, etc.)  what you're seeing the C++-side of the implementation of each method.  I'm hoping there's a similar paradigm available for Java you can take advantage of.

Apologies if this is all super well known to you guys (i know about zero about Java) and I wasted your time pointing out the obvious.  I just got a little nervous hearing about maybe making a Java map to reflect the C++ one.

Reply:
This is how everything is implemented right now.

For example https://github.com/KarthikRIyer/OpenTimelineIO/blob/fd5a269bfedf31c0b152fb83f86b641f95a2ce55/src/java-opentimelineio/src/main/java/io/opentimeline/opentimelineio/AnyDictionary.java#L101

This is similar to the get item method. If you see other functions with the native keyword in the declaration they will have a cpp implementation. 

@margant
I've tried to implement commonly used functions in java HashMap or ArrayList. We wouldn't be able to use all the stuff in the Java collections API but it does work like a map or list. For example We wouldn't be ale to use the for each loop. Is there any way we could get such stuff working?

Reply:
Agree with @davidbaraff we should not have to clone entire collection on a get operation, it can get costly with unnecessary memory allocations.

To preserve the "feel" of Java we should definitely try extending from
https://docs.oracle.com/javase/8/docs/api/java/util/AbstractCollection.html or implement https://docs.oracle.com/javase/8/docs/api/java/util/Collection.html interface. Implementing  [Iterable](https://docs.oracle.com/javase/8/docs/api/java/lang/Iterable.html) can get you the for-each with less work then above, but Collection has nicer elements such as size and others that make it more useful. If we do nothing, I would at least strongly consider having a toCollection() method that returns a more "pure" Java collection object from our custom ones.

Reply:
I'll look into extending hese classes.